pkg/review/job.go

```diff
@@ -81,61 +75,78 @@ func ExecuteJob(ctx context.Context, cfg *Config) error {
 	}
 
 	// Step 2: Get review status information for each commit.
-	// Fan out the work of processing all of the commits that were found
-	for _, commit := range commits {
-		if err := commitPool.Do(ctx, func() (*CommitReviewStatus, error) {
-			result := processCommit(ctx, *commit, gitHubClient)
-			return result, nil
-		}); err != nil {
-			return fmt.Errorf("failed to submit job to worker pool: %w", err)
-		}
-	}
-	// When all of the workers are complete, extract the result values
-	results, err := commitPool.Done(ctx)
+	commitReviewStatuses, err := pooledTransform(ctx, commits,
+		func(commit *Commit) (*CommitReviewStatus, error) {
+			return processCommit(ctx, gitHubClient, commit), nil
+		},
+	)
 	if err != nil {
 		return fmt.Errorf("failed to process commits: %w", err)
 	}
-	statuses := make([]*CommitReviewStatus, 0, len(results))
-	for _, v := range results {
-		// Statuses that were failed to be retrieved will be nil
-		if v.Value != nil {
-			statuses = append(statuses, v.Value)
-		}
+	// statuses that should not be further processed will be nil, so we should exclude them
+	commitReviewStatuses = removeNil(commitReviewStatuses)
+
+	// Step 3: Look up break glass issue if necessary and tag the review status with it if found.
+	fetcher := &BigQueryBreakGlassIssueFetcher{
+		client: bqClient,
+	}
+	taggedReviewStatuses, err := pooledTransform(ctx, commitReviewStatuses,
+		func(status *CommitReviewStatus) (*CommitReviewStatus, error) {
+			return processReviewStatus(ctx, fetcher, cfg, status), nil
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to process commit review statuses: %w", err)
+	}
+	// statuses that should not be further processed will be nil
+	taggedReviewStatuses = removeNil(taggedReviewStatuses)
+
+	// Step 4: Write the commit review status information to BigQuery.
+	if err := bq.Write[CommitReviewStatus](ctx, bqClient, cfg.CommitReviewStatusTableID, taggedReviewStatuses); err != nil {
+		return fmt.Errorf("failed to write commit review statuses to bigquery: %w", err)
 	}
 
-	// Step 3: Look up break glass issue if necessary.
-	statusPool := workerpool.New[*CommitReviewStatus](&workerpool.Config{
+	return nil
+}
+
+// pooledTransform transforms each input element of type T into an element of type V using the given transform function.
+// The transform is fanned out using a worker pool so that each input element may be processed asynchronously from the
+// others.
+func pooledTransform[T, V any](ctx context.Context, elements []T, transform func(T) (V, error)) ([]V, error) {
```

Comment:
I don't understand the desire to abstract this, and the abstraction is not idiomatic. Go does not typically do functional programming patterns, and it's usually a substantial performance overhead over inlining.

Reply:
There was a bug where the wrong slice was being referenced due to the worker pool code logic being copy/pasted. 

The original code with the bug is here:
https://github.com/abcxyz/github-metrics-aggregator/blob/6caa020e41f833bf7e3cf72d785f7f6742a03f8c/pkg/review/job.go#L132

After the looking at the worker pool code it was clear that there was a lot of boilerplate going on in setting up the pool and accumulating the results. So i figured I would extract the worker pool code into a function in order to avoid this type of error from occurring again. 

I'm not sure how much a performance hit there really is in this case. When the job is invoked, the `pooledTransform` function will be called exactly twice. If this were say a function that would be called many times I could see the argument for inlining being better for performance (and would likely agree that it was the better choice).